packages/material-ui/src/styles/muiStyled.js

```diff
@@ -0,0 +1,61 @@
+import styled from '@material-ui/styled-engine';
+import { propsToClassKey } from '@material-ui/styles';
+
+export const getStyleOverrides = (name, theme) => {
+  let styleOverrides = {};
+
+  if (
+    theme &&
+    theme.components &&
+    theme.components[name] &&
+    theme.components[name].styleOverrides
+  ) {
+    styleOverrides = theme.components[name].styleOverrides;
+  }
+
+  return styleOverrides;
+};
+
+export const getVariantStyles = (name, theme) => {
+  let variants = [];
+  if (theme && theme.components && theme.components[name] && theme.components[name].variants) {
+    variants = theme.components[name].variants;
+  }
+
+  const variantsStyles = {};
+
+  variants.forEach((definition) => {
+    const key = propsToClassKey(definition.props);
+    variantsStyles[key] = definition.style;
+  });
+
+  return variantsStyles;
+};
+
+const shouldForwardProp = (prop) => prop !== 'state' && prop !== 'theme';
+
+const muiStyled = (el, params, muiConfig) => {
+  const result = styled(el, { shouldForwardProp, ...params });
```

Comment:
`styled-components` does not play well when the `styled` functionality is invoked in another function :(